homeassistant/components/aprilaire/coordinator.py

```diff
@@ -0,0 +1,180 @@
+"""The Aprilaire coordinator."""
+
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from logging import Logger
+from typing import Any
+
+import pyaprilaire.client
+from pyaprilaire.const import MODELS, Attribute, FunctionalDomain
+
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.device_registry as dr
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN
+
+RECONNECT_INTERVAL = 60 * 60
+RETRY_CONNECTION_INTERVAL = 10
+
+
+class AprilaireCoordinator(DataUpdateCoordinator):
+    """Coordinator for interacting with the thermostat."""
+
+    def __init__(
+        self, hass: HomeAssistant, host: str, port: int, logger: Logger
+    ) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass,
+            logger,
+            name=DOMAIN,
+        )
+
+        self.data: dict[str, Any] = {}
+
+        self.client = pyaprilaire.client.AprilaireClient(
+            host,
+            port,
+            self.async_set_updated_data,
+            self.logger,
+            RECONNECT_INTERVAL,
+            RETRY_CONNECTION_INTERVAL,
+        )
+
+    def async_set_updated_data(self, data: Any) -> None:
+        """Manually update data, notify listeners and reset refresh interval."""
+
+        old_device_info = self.create_device_info(self.data)
+
+        if self.data is not None:
+            data = self.data | data
+
+        super().async_set_updated_data(data)
+
+        new_device_info = self.create_device_info(data)
+
+        if (
+            old_device_info is not None
+            and new_device_info is not None
+            and old_device_info != new_device_info
+        ):
+            device_registry = dr.async_get(self.hass)
+
+            device = device_registry.async_get_device(old_device_info["identifiers"])
+
+            if device is not None:
+                new_device_info.pop("identifiers", None)
+                new_device_info.pop("connections", None)
+
+                device_registry.async_update_device(
+                    device_id=device.id, **new_device_info  # type: ignore[misc]
+                )
+
+    async def start_listen(self):
+        """Start listening for data."""
+        await self.client.start_listen()
+
+    def stop_listen(self):
+        """Stop listening for data."""
+        self.client.stop_listen()
+
+    async def wait_for_ready(
```

Comment:
I am no expert on this, but maybe you can create a task for everything you wait for and use it with asyncio.gather. But I am not sure if that is something good to use here. But just putting it out here

Reply:
@joostlek It is _possible_ to, but the complication here is that the connection doesn't work in a request/response manner like you might expect. Packets are sent to the thermostat with no acknowledgement or response, and at some point in the future the thermostat sends an altogether separate packet with the data. Additionally, the thermostat is very sensitive to receiving too many packets at once, which can cause it to ignore some, send errors, or in a worst-case scenario lock up entirely. So, sending these serially here is in my opinion the best approach, though it definitely is not the cleanest from a code perspective.